ocaml/xapi/importexport.ml

```diff
@@ -16,52 +16,73 @@
 *)
 
 (** Represents a database record (the reference gets converted to a small string) *)
-type obj = { cls: string; id: string; snapshot: XMLRPC.xmlrpc }
+type obj = {
+  cls: string [@key "class"];
+  id: string;
+  snapshot: Rpc.t
+} [@@deriving rpc]
 
-(** Version information attached to each export and checked on import *)
-type version =
-  { hostname: string;
-    date: string;
-    product_version: string;
-    product_brand: string;
-    build_number: string;
-    xapi_vsn_major: int;
-    xapi_vsn_minor: int;
-    export_vsn: int; (* 0 if missing, indicates eg whether to expect sha1sums in the stream *)
-  }
+let string_of_obj x = x.cls ^ "  " ^ x.id
 
-(** An exported VM has a header record: *)
-type header =
-  { version: version;
-    objects: obj list }
+(** Version information attached to each export and checked on import *)
+type version = {
+  hostname: string;
+  date: string;
+  product_version: string;
+  product_brand: string;
+  build_number: string;
+  xapi_vsn_major: int;
+  xapi_vsn_minor: int;
+  export_vsn: int; (* 0 if missing, indicates eg whether to expect sha1sums in the stream *)
+}
 
-exception Version_mismatch of string
+let rpc_of_version x =
+  let open Xapi_globs in
+  Rpc.Dict(
+    [ _hostname,        Rpc.String(x.hostname)
+    ; _date,            Rpc.String(x.date)
+    ; _product_version, Rpc.String(x.product_version)
+    ; _product_brand,   Rpc.String(x.product_brand)
+    ; _build_number,    Rpc.String(x.build_number)
+    ; _xapi_major,      Rpc.Int(Int64.of_int Xapi_globs.version_major)
+    ; _xapi_minor,      Rpc.Int(Int64.of_int Xapi_globs.version_minor)
+    ; _export_vsn,      Rpc.Int(Int64.of_int Xapi_globs.export_vsn)
+    ])
 
-module D=Debug.Make(struct let name="importexport" end)
-open D
+exception Failure of string
 
 let find kvpairs where x =
   if not(List.mem_assoc x kvpairs)
   then raise (Failure (Printf.sprintf "Failed to find key '%s' in %s" x where))
   else List.assoc x kvpairs
 
-let string_of_obj x = x.cls ^ "  " ^ x.id
+[@@@warning "-8"]
+let version_of_rpc = function
+  | Rpc.Dict(map) ->
+    let find = find map "version data" in
+    let open Xapi_globs in
+    { hostname        = Rpc.string_of_rpc (find _hostname)
+    ; date            = Rpc.string_of_rpc (find _date)
+    ; product_version = Rpc.string_of_rpc (find _product_version)
+    ; product_brand   = Rpc.string_of_rpc (find _product_brand)
+    ; build_number    = Rpc.string_of_rpc (find _build_number)
+    ; xapi_vsn_major  = Rpc.int_of_rpc (find _xapi_major)
+    ; xapi_vsn_minor  = Rpc.int_of_rpc (find _xapi_minor)
+    ; export_vsn      = try Rpc.int_of_rpc (find _export_vsn) with _ -> 0
+    }
+  | rpc -> raise (Failure(Printf.sprintf "version_of_rpc: malformed RPC %s" (Jsonrpc.to_string rpc)))
```

Comment:
This is a minor point, but I'd use `Rpc.to_string` here, because we're dealing with a generic Rpc structure here.

Reply:
Faor enough. I uesd json only because I find it more readable but it is more natural to use Rpc.to_string. I’ll change it